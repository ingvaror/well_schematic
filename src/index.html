<!DOCTYPE html>
<html lang="en" xmlns="http://www.w3.org/1999/xhtml" style="width:100%;height:100%;">
<head>
    <meta charset="utf-8" />
    <meta name="viewport" content="user-scalable=no, width=device-width, height=device-height, initial-scale=1, maximum-scale=1" />
    <title>Well Schematic</title>
</head>
  <body height="100%">
      <div id="root" height="100%"></div>
  </body>
</html>
