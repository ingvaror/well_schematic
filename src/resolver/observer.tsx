/** Observable pattern from https://refactoring.guru/design-patterns/observer/typescript/example 
 It's little modified to provide Generic Data from udpate function (not Subject) and it's more observable than observer
 */

/**
 * The Observer interface declares the update method, used by subjects.
 */
export interface Listener<DataType> {
    // Receive update from observable.
    update(data: DataType): void
}

/**
 * The Observable owns some important state and notifies observers when the state
 * changes.
 */
export class Observable<DataType> {
    /**
    * @type {number} For the sake of simplicity, the Subject's state, essential
    * to all subscribers, is stored in this variable.
    */
    public data: DataType

    /**
     * @type {Listener[]} List of subscribers. In real life, the list of
     * subscribers can be stored more comprehensively (categorized by event
     * type, etc.).
     */
    private listeners: Listener<DataType>[] = []

    /**
     * The subscription management methods.
     */
    public attach(listener: Listener<DataType>): void {
        const isExist = this.listeners.includes(listener)
        if (!isExist)
            this.listeners.push(listener)
    }

    public detach(listener: Listener<DataType>): void {
        const observerIndex = this.listeners.indexOf(listener)
        if (observerIndex === -1) {
            alert('Observable: Nonexistent observer.')
        }

        this.listeners.splice(observerIndex, 1)
    }

    /**
     * Trigger an update in each subscriber.
     */
    public notify(newData: DataType): void {
        this.data = newData
        for (const listener of this.listeners) {
            listener.update(newData)
        }
    }
}