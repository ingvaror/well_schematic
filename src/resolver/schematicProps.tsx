
export enum SchematicDirection {
    RIGHT_HAND_SIDE,
    LEFT_HAND_SIDE
}

export class SchematicPoint {
    X: number
    Y: number
    TVD: number
    MD: number
    MDH: number
    sinInclination: number
    cosInclination: number
    Rwell: number
}

export interface WellTrajectory {
    viewDiameterWell: number
	radiusFactor: number
    points: Array<SchematicPoint>
}

export interface SchematicLayer {
    key: string
    layerName: string
    description: string
    completionName: string
    inner: number
    outer: number
    gravelInAnnulus: boolean
    packerLeaking: boolean
}

export interface SchematicCompletionDevice {
    completionName: string
    layerName: string
    outer: number
    x1: number
    y1: number
    rotate: number
    scaleWidth: number
    scaleHeight: number
}

export interface SchematicCompletionPipe {
    completionName: string
    layerName: string
}

export interface SchematicCompletion {
    key: string
    item: string
    userNotes: string
	MD: number
	TVD: number
	length: number
    viewDiameterWell: number
    layers?: Array<SchematicLayer>
    points: Array<SchematicPoint>
    devices: Array<SchematicCompletionDevice>
    pipes: Array<SchematicCompletionPipe>
}

export interface PressureSensor {
	key: string
	MD: number
    radiuses: Array<number>
}

export interface SchematicFluidComposition {
    layerName: string
    Q: number
    Qoil: number
    Qgas: number
    Qwater: number
    inner: number
    outer: number
    MD: number
}

export interface SchematicSimulationData {
    layerName: string
    value: number
    inner: number
    outer: number
    MD: number
}

export interface SchematicSimulationResults {
    num: number    //here can be time or Qoil for IPR
    title: string
    unitType: string
    influxRate: Array<SchematicSimulationData>
    flowRate: Array<SchematicSimulationData>
    fractions: Array<SchematicFluidComposition>
    pressure: Array<SchematicSimulationData>
    temperature: Array<SchematicSimulationData>
}

export interface RangesSimulationResults
{
	minValue: number
	maxValue: number
}

export interface UnitVariant {
    name: string
    factor: number
	offset: number
}

export interface SchematicViewOptions {
    modeView: string
}

export interface ScheduleParameters {
    num?: number
    colorized: string
}

export const SCHEMATIC_RATES = "Flow rate"
export const SCHEMATIC_FLUID_COMPOSITION = "Phase flow fractions"
export const SCHEMATIC_PRESSURE = "Pressure"
export const SCHEMATIC_TEMPERATURE = "Temperature"
export const SCHEMATIC_NOTHING = "Nothing"
export const SCHEMATIC_INFLUX_RATE = "Influx rate"

export const SCHEMATIC_UNIT_LENGTH = "length"
export const SCHEMATIC_UNIT_DIAMETER_LENGTH = "length-diameters"
export const SCHEMATIC_UNIT_FLOWRATE = "flowrate-reservoir"
export const SCHEMATIC_UNIT_FLUXRATE = "flux-reservoir"
export const SCHEMATIC_UNIT_PRESSURE = "pressure"
export const SCHEMATIC_UNIT_TEMPERATURE = "temperature"

export const NETOOL_WELLSCHEMATIC_VERTICAL = "NETOOL_WELLSCHEMATIC_VERTICAL"
export const NETOOL_WELLSCHEMATIC_DEVIATION = "NETOOL_WELLSCHEMATIC_DEVIATION"
export const NETOOL_WELLSCHEMATIC_DEVIATION_WITH_AXES = "NETOOL_WELLSCHEMATIC_DEVIATION_WITH_AXES"

export enum CompletionType {
    BLANK_PIPE_HTML = "BLANK_PIPE_HTML",
    CEMENTED_BLANK_PIPE_HTML = "CEMENTED_BLANK_PIPE_HTML",
    ICD_HTML = "ICD_HTML",
    ICV_HTML = "ICV_HTML",
    JUNCTION_HTML = "JUNCTION_HTML",
    MISSED_COMPLETION_HTML = "MISSED_COMPLETION_HTML",
    
    PACKER_HTML = "PACKER_HTML",

    CHOKE_HTML = "CHOKE_HTML",
    PLUG_HTML = "PLUG_HTML",
    PUMP_ESP_HTML = "PUMP_ESP_HTML",
    PUMP_PCP_HTML = "PUMP_PCP_HTML",
    PUMP_SRP_HTML = "PUMP_SRP_HTML",

    OPEN_HOLE_HTML = "OPEN_HOLE_HTML",
    PERF_CEMENTED_HTML = "PERF_CEMENTED_HTML",
    SCREEN_HTML = "SCREEN_HTML",
    SLOTTED_HTML = "SLOTTED_HTML",

    SECTION_ICD_HTML = "SECTION_ICD_HTML",
    SECTION_ICV_HTML = "SECTION_ICV_HTML",
    SECTION_PACKER_HTML = "SECTION_PACKER_HTML"
}
