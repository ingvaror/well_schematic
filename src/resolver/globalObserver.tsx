import { Observable } from "./observer"


export class GlobalObserver {
    private static instance: GlobalObserver

    public static getInstance(): GlobalObserver {
        if (this.instance == null) {
            this.instance = new GlobalObserver()
            this.instance.observableShowTable.data = true;
        }
        return this.instance
    }

    private scrolling: boolean = false;

    public observableResolverInit: Observable<boolean> = new Observable<boolean>()

    public observableZoomIn: Observable<void> = new Observable<void>()
    public observableZoomOut: Observable<void> = new Observable<void>()
    public observableZoomReset: Observable<void> = new Observable<void>()
    public observableShowTable: Observable<boolean> = new Observable<boolean>()
    public observableScrollingSVG: Observable<number> = new Observable<number>()
    public observableScrollingTable: Observable<number> = new Observable<number>()
    public isScrolling(): boolean {
        return this.scrolling;
    }
    public setScrolling(_scrolling: boolean) {
        this.scrolling = _scrolling;
    }

    public observableSetCompletionFromTable: Observable<string> = new Observable<string>()
    public observableSetCompletionFromSVG: Observable<string> = new Observable<string>()
    public observableUnsetCompletionFromTable: Observable<string> = new Observable<string>()
    public observableUnsetCompletionFromSVG: Observable<string> = new Observable<string>()
}