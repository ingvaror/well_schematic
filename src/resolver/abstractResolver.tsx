import { ZoomProps } from "../svg/zoomProps"
import { Observable } from "./observer"
import { PressureSensor, ScheduleParameters, SchematicCompletion, SchematicSimulationResults, UnitVariant, WellTrajectory } from "./schematicProps"
import { WellSchematicData } from "./wellSchematicData"

export interface ObservableScheduleType {
    simulationResults: Array<SchematicSimulationResults>
    currentschedule: ScheduleParameters
}

export interface ObservablePanZoomType {
    completions: Array<SchematicCompletion>
    pressureSensors: Array<PressureSensor>
    simulationResults: Array<SchematicSimulationResults>
}

export interface ObservableViewDiameterWellType {
    completions: Array<SchematicCompletion>
    pressureSensors: Array<PressureSensor>
    simulationResults: Array<SchematicSimulationResults>
    wellTrajectory: WellTrajectory
}

export interface AbstractResolver {
    /**
     * Method to check that initialization of resolver is completed
     * */
    isInitiated(): boolean

    /** App call this method after initialization to set current lateral */
    getFirstLateral()
    /** completionTable and colorizedScale call this method to get units which is saved in WEB_Schematic_View */
    getUnits()

    /**
     * Call this method when title of lateral is changed
     * @param titleLateral Title from combobox
     */
    onUpdateTitleLateral(titleLateral: string)
    // Callback when new data has received. That's all well schematic data
    observableLateral: Observable<WellSchematicData>

    /**
     * @param scheduleParameters It includes num which is number of schedule (days, oil rates, etc.) 
     * and visualized simulation parameter (temperature, pressure, etc.)
     */
    onUpdateSchedule(scheduleParameters: ScheduleParameters)
    // Callback when new data has received. It't only Simulation Results from selected lateral
    observableSchedule: Observable<ObservableScheduleType>

    /**
     * Call this method after Pan&Zoom in 2D view or scrolling in vertical view
     * @param zoomProps Client visible field
     */
    onUpdatePanZoom(zoomProps: ZoomProps)
    // Callback when new data has received
    observablePanZoom: Observable<ObservablePanZoomType>

    ///** Call this method after update units */
    // Callback when new units has received
    observableUnits: Observable<Map<string, UnitVariant>>

    /**
     * 
     * Call this method when view diameter well is changed by slider
     * @param viewDiameterWell new view diameter
     */
    onUpdateViewDiameterWell(viewDiameterWell: number)
    // Callback when new data has received from server about completions and well trajectory
    observableViewDiameterWell: Observable<ObservableViewDiameterWellType>
}