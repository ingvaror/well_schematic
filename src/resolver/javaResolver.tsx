import { ZoomProps } from "../svg/zoomProps"
import { AbstractResolver, ObservablePanZoomType, ObservableScheduleType, ObservableViewDiameterWellType } from "./abstractResolver"
import { Observable } from "./observer"
import { PressureSensor, ScheduleParameters, SchematicCompletion, SchematicSimulationResults, UnitVariant } from "./schematicProps"
import { WellSchematicData } from "./wellSchematicData"

interface JavaConnector 
{
    getFirstLateral: () => void
    getUnits: () => void
    //setClipBoardData: (value: string) => void
    copyToClipboard: (text: string, content_type : string) => void
    onUpdateTitleLateral: (title: string) => void
    onUpdateSchedule: (num: number, colorized: string) => void
    onUpdatePanZoom: (leftUpCornerX: number, leftUpCornerY: number, rightDownCornerX: number, rightDownCornerY: number, realZoom: number) => void
    onUpdateViewDiameterWell: (viewDiameterWell: number) => void
}

export class JavaResolver implements AbstractResolver {
    // Use to call functions Java (uses by)
    private javaConnector: JavaConnector

    private updateJavaConnector() {
        this.javaConnector = (window as any).javaConnector
    }

    constructor() {
        this.javaConnector = (window as any).javaConnector
    }

    public isInitiated(): boolean {
        if (this.javaConnector != null)
            return true
        else {
            this.javaConnector = (window as any).javaConnector
            if (this.javaConnector != null)
                return true
            else
                return false
        }
    }

    public getFirstLateral() {
        if (this.javaConnector != null)
            this.javaConnector.getFirstLateral()
    }

    public getUnits() {
        if (this.javaConnector != null)
            this.javaConnector.getUnits()
    }

    public onUpdateUnits(newUnits: Map<string, UnitVariant>) {
        this.onUpdateUnits(newUnits)
    }

    public onUpdateTitleLateral(titleLateral: string) {
        this.updateJavaConnector()
        if (this.javaConnector != null)
            this.javaConnector.onUpdateTitleLateral(titleLateral)
    }
    public observableLateral: Observable<WellSchematicData> = new Observable<WellSchematicData>()

    public onUpdateSchedule(scheduleParameters: ScheduleParameters) {
        this.updateJavaConnector()
        if (this.javaConnector != null)
            this.javaConnector.onUpdateSchedule(scheduleParameters.num, scheduleParameters.colorized)
    }
    public observableSchedule: Observable<ObservableScheduleType> = new Observable<ObservableScheduleType>()

    public onUpdatePanZoom(zoomProps: ZoomProps) {
        this.updateJavaConnector()
        if (this.javaConnector != null)
            this.javaConnector.onUpdatePanZoom(zoomProps.leftUpCornerX, zoomProps.leftUpCornerY, zoomProps.rightDownCornerX, zoomProps.rightDownCornerY, zoomProps.realZoom)
    }
    public observablePanZoom: Observable<ObservablePanZoomType> = new Observable<ObservablePanZoomType>()

    public observableUnits: Observable<Map<string, UnitVariant>> = new Observable<Map<string, UnitVariant>>()

    public onUpdateViewDiameterWell(viewDiameterWell: number) {
        this.updateJavaConnector()
        if (this.javaConnector != null) {
            this.javaConnector.onUpdateViewDiameterWell(viewDiameterWell)
        }
    }

    public observableViewDiameterWell: Observable<ObservableViewDiameterWellType> = new Observable<ObservableViewDiameterWellType>()

    public getJavaConnector(): JavaConnector {
        this.updateJavaConnector()
        return this.javaConnector
    }

}
