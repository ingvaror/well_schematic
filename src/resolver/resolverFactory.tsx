import { AbstractResolver } from "./abstractResolver"
import { GlobalObserver } from "./globalObserver"
import { JavaResolver } from "./javaResolver"
import { StandaloneResolver } from "./standaloneResolver"
import { TestResolver } from "./testResolver"
import { WebsocketResolver } from "./websocketResolver"

export class ResolverFactory {
    private static instance: AbstractResolver

    public static readonly JAVA_RESOLVER = "JAVA_RESOLVER"

    public static readonly WEBSOCKET_RESOLVER = "WEBSOCKET_RESOLVER"

    public static readonly STANDALONE_RESOLVER = "STANDALONE_RESOLVER"

    public static readonly TEST_RESOLVER = "TEST_RESOLVER"

    public static initResolver(type: string) {
        if (this.instance != null)
            return;
        switch (type) {
            case ResolverFactory.JAVA_RESOLVER:
                this.instance = new JavaResolver()
                break
            case ResolverFactory.WEBSOCKET_RESOLVER:
                this.instance = new WebsocketResolver()
                break
            case ResolverFactory.STANDALONE_RESOLVER:
                this.instance = new StandaloneResolver()
                break
            case ResolverFactory.TEST_RESOLVER:
                this.instance = new TestResolver()
                break
            default:
                alert("Undefined remote resolver. Should use " + ResolverFactory.WEBSOCKET_RESOLVER + " or " + ResolverFactory.JAVA_RESOLVER)
        }
        var interval = setInterval(() => {
            if (this.instance.isInitiated()) {
                GlobalObserver.getInstance().observableResolverInit.notify(true)
                clearInterval(interval)
            }
        }, 200)
    }

    public static getResolver(): AbstractResolver {
        return this.instance
    }

}