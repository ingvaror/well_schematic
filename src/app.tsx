import * as React from "react"

import { WellSchematicData } from "./resolver/wellSchematicData"
import { NETOOL_WELLSCHEMATIC_VERTICAL, UnitVariant, } from "./resolver/schematicProps"

import { ResolverFactory } from "./resolver/resolverFactory"
import { JavaResolver } from "./resolver/javaResolver"
import { Listener } from "./resolver/observer"
import { ObservablePanZoomType, ObservableScheduleType, ObservableViewDiameterWellType } from "./resolver/abstractResolver"
import './app.css'
import { GlobalObserver } from "./resolver/globalObserver"
import { DivSVG } from "./div/divSVG"
import { DivTable } from "./div/divTable"
import { WebsocketResolver } from "./resolver/websocketResolver"
import { ZoomProps } from "./svg/zoomProps"

// Global functions for Java Desktop. WEB_Schematic_View.JavaConnectorSchematic executes it.

// Igor does not want units updates right after the document is loaded,
// he wants to delay until the end of React.useEffect();
// It's much easier to do it here than screw all the Java logic;
//
var can_update_units = false;

// Java now checks for presence of this function => long name to make sure
// it won't accidentally find it in foreign html.
//
(window as any).netool_updateUnits = (newUnits: Map<string, UnitVariant>) => 
{
	if (can_update_units)
	{
		if (ResolverFactory.getResolver() instanceof JavaResolver)
			(ResolverFactory.getResolver() as JavaResolver).observableUnits.notify(newUnits);
	}
}

(window as any).updateLateral = (data: WellSchematicData) => {
	if (ResolverFactory.getResolver() instanceof JavaResolver)
		(ResolverFactory.getResolver() as JavaResolver).observableLateral.notify(data)
}

(window as any).updateSchedule = (data: ObservableScheduleType) => {
	if (ResolverFactory.getResolver() instanceof JavaResolver)
		(ResolverFactory.getResolver() as JavaResolver).observableSchedule.notify(data)
}

(window as any).updatePanZoom = (data: ObservablePanZoomType) => {
	if (ResolverFactory.getResolver() instanceof JavaResolver)
		(ResolverFactory.getResolver() as JavaResolver).observablePanZoom.notify(data)
}

(window as any).updateViewDiameterWell = (data: ObservableViewDiameterWellType) => {
	if (ResolverFactory.getResolver() instanceof JavaResolver)
		(ResolverFactory.getResolver() as JavaResolver).observableViewDiameterWell.notify(data)
}

const onSetResolverInit: Listener<boolean> = { update: (_: boolean) => void {} }
const onChangeWellSchematicData: Listener<WellSchematicData> = { update: (_: WellSchematicData) => void {} }
const onChangeSchedule: Listener<ObservableScheduleType> = { update: () => void {} }
const onChangeUnits: Listener<Map<string, UnitVariant>> = { update: () => void {} }
const onChangeViewDiameterWell: Listener<ObservableViewDiameterWellType> = { update: (_: ObservableViewDiameterWellType) => void {} }
const onSetShowTable: Listener<boolean> = { update: (_: boolean) => void {} }

var isUpdatedUnits = false

export const App = () => {

	ResolverFactory.initResolver(ResolverFactory.WEBSOCKET_RESOLVER)

	const [wellSchematicData, setWellSchematicData] = React.useState<WellSchematicData>(null)

	const [resolverInit, setResolverInit] = React.useState<boolean>(false)

	const [showTable, setShowTable] = React.useState<boolean>(GlobalObserver.getInstance().observableShowTable.data)

	const currentLateral = wellSchematicData != null ? wellSchematicData.laterals.find((lateral) => { return lateral.title == wellSchematicData.currentTitleLateral }) : null

	const getWidthLeftPanel = (): number => {
		if (wellSchematicData != null && wellSchematicData.viewOptions.modeView == NETOOL_WELLSCHEMATIC_VERTICAL)
			return 15
		else if (!showTable)
			return 100
		else
			return 50
	}

	/** After init App */

	React.useEffect(() => {
		ResolverFactory.getResolver().observableLateral.attach(onChangeWellSchematicData)
		ResolverFactory.getResolver().observableViewDiameterWell.attach(onChangeViewDiameterWell)
		ResolverFactory.getResolver().observableSchedule.attach(onChangeSchedule)
		ResolverFactory.getResolver().observableUnits.attach(onChangeUnits)
		GlobalObserver.getInstance().observableResolverInit.attach(onSetResolverInit)
		GlobalObserver.getInstance().observableShowTable.attach(onSetShowTable)

		onChangeWellSchematicData.update = (newData: WellSchematicData) => {
			if (wellSchematicData != null) {
				newData.units = wellSchematicData.units
			}
			setWellSchematicData(newData)
		}
		onChangeViewDiameterWell.update = (newData: ObservableViewDiameterWellType) => {
			wellSchematicData.laterals
				.filter((lateral) => { return lateral.title == currentLateral.title })
				.forEach((lateral) => {
					lateral.completions = newData.completions
					lateral.pressureSensors = newData.pressureSensors
					lateral.simulationResults = newData.simulationResults
					lateral.wellTrajectory = newData.wellTrajectory
				})
			setWellSchematicData(wellSchematicData)
		}
		onChangeSchedule.update = (newData: ObservableScheduleType) => {
			wellSchematicData.currentschedule = newData.currentschedule
			wellSchematicData.laterals
				.filter((lateral) => { return lateral.title == currentLateral.title })
				.forEach((lateral) => {
					lateral.simulationResults = newData.simulationResults
				})
			setWellSchematicData(wellSchematicData)
		}
		onChangeUnits.update = (newData: Map<string, UnitVariant>) => {
			wellSchematicData.units = newData
			setWellSchematicData({ ...wellSchematicData })
		}

		onSetResolverInit.update = (resolverInit: boolean) => { setResolverInit(resolverInit) }
		onSetShowTable.update = (showTable: boolean) => { setShowTable(showTable) }

		if (wellSchematicData != null && wellSchematicData.viewOptions != null &&
			wellSchematicData.viewOptions.modeView == NETOOL_WELLSCHEMATIC_VERTICAL &&
			(ResolverFactory.getResolver() instanceof JavaResolver || ResolverFactory.getResolver() instanceof WebsocketResolver)) {
			const zoomProps: ZoomProps = {
				leftUpCornerX: -window.innerWidth,
				leftUpCornerY: 0,
				rightDownCornerX: window.innerWidth,
				rightDownCornerY: window.innerHeight,
				realZoom: 1.0
			}
			ResolverFactory.getResolver().onUpdatePanZoom(zoomProps)
		}
		if (resolverInit && wellSchematicData == null) {
			ResolverFactory.getResolver().getFirstLateral()
		}
		
		can_update_units = true;
		
		if (resolverInit && wellSchematicData != null && !isUpdatedUnits &&
			(ResolverFactory.getResolver() instanceof JavaResolver || ResolverFactory.getResolver() instanceof WebsocketResolver)) {
			// Only for Java UI. If someone changes units while downloading schematic React App
			isUpdatedUnits = true;
			ResolverFactory.getResolver().getUnits()
		}
	})

	if (resolverInit && wellSchematicData != null) {
		if (wellSchematicData.laterals == null || wellSchematicData.laterals.length == 0) {
			return <p className="warningText"><br /><br /><br /><big>Well schematic will be shown after well trajectory and well segments are created.</big></p>
		}
		else {
			if (showTable || wellSchematicData != null && wellSchematicData.viewOptions != null &&
				wellSchematicData.viewOptions.modeView == NETOOL_WELLSCHEMATIC_VERTICAL) {
				return <>
					<DivSVG currentLateral={currentLateral} wellSchematicData={wellSchematicData} widthLeftPanel={getWidthLeftPanel()} />
					<DivTable currentLateral={currentLateral} wellSchematicData={wellSchematicData} widthLeftPanel={getWidthLeftPanel()} />
				</>
			} else {
				return <DivSVG currentLateral={currentLateral} wellSchematicData={wellSchematicData} widthLeftPanel={getWidthLeftPanel()} />
			}
		}
	} else {
		return <p className="warningText"><br /><br /><br /><big>Loading well schematic...</big></p>
	}
}
