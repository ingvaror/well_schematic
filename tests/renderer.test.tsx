import * as React from "react";
import { act } from "react-dom/test-utils";
import * as ReactDOM from 'react-dom';
import { App } from "../src/app"
import { ResolverFactory } from "../src/resolver/resolverFactory";
import { GlobalObserver } from "../src/resolver/globalObserver";

let container;

beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
});

afterEach(() => {
    document.body.removeChild(container);
    container = null;
});

it('Initial message', () => {
    ResolverFactory.initResolver(ResolverFactory.TEST_RESOLVER)
    act(() => {
        ReactDOM.render(<App />, container);
    });
    act(() => { GlobalObserver.getInstance().observableResolverInit.notify(true) })
    expect(container).toMatchSnapshot()
})

it('Empty case', () => {
    ResolverFactory.initResolver(ResolverFactory.TEST_RESOLVER)
    act(() => {
        ReactDOM.render(<App />, container);
    });
    act(() => { GlobalObserver.getInstance().observableResolverInit.notify(true) })
    act(() => { ResolverFactory.getResolver().observableLateral.notify({ laterals: [], units: null, viewOptions: null, currentschedule: null, currentTitleLateral: null, availableColorized: null, rangessimulationresults: null }) })
    expect(container).toMatchSnapshot()
})
