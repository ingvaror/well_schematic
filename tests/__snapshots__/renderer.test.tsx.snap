// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Empty case 1`] = `
<div>
  <p
    class="warningText"
  >
    <br />
    <br />
    <br />
    <big>
      Well schematic will be shown after well trajectory and well segments are created.
    </big>
  </p>
</div>
`;

exports[`Initial message 1`] = `
<div>
  <p
    class="warningText"
  >
    <br />
    <br />
    <br />
    <big>
      Loading well schematic...
    </big>
  </p>
</div>
`;
